### PR TITLE
Added dimension fields for all datastreams of Oracle Integration

### PR DESCRIPTION
**Pull request by agithomas**
Type of change
- Enhancement

## What does this PR do?

Adding dimension fields for all datastream under Oracle Integration package to take the advantage of TSDB

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [ ]

## Related issues

-

### CODE REVIEW

File: packages/oracle/data_stream/memory/fields/agent.yml [hunk 1]
```diff
@@ -0,0 +1,213 @@
+- name: cloud
```
Comment (ruflin) [inline packages/oracle/data_stream/memory/fields/agent.yml:1]: Was this file missing before? As tehse are all (most?) ECS field, should we just reference it?

File: packages/oracle/data_stream/memory/fields/base-fields.yml [hunk 2]
```diff
@@ -6,6 +6,7 @@
   description: Data stream dataset.
 - name: data_stream.namespace
   type: constant_keyword
+  dimension: true
```
Comment (ruflin) [inline packages/oracle/data_stream/memory/fields/base-fields.yml:9]: See my comment in the issue. It is not clear to me why this should be a dimension.

File: packages/oracle/data_stream/performance/fields/fields.yml [hunk 3]
```diff
@@ -156,4 +164,5 @@
           description: Number of times waits of the class occurred for the session.
         - name: wait_class
           type: keyword
+          dimension: true
```
Comment (ruflin) [inline packages/oracle/data_stream/performance/fields/fields.yml:167]: What is the total number of dimensions we are hitting in performance dataset?

File: packages/oracle/data_stream/tablespace/fields/fields.yml [hunk 4]
```diff
@@ -41,11 +43,13 @@
 
         - name: status
           type: keyword
+          dimension: true
           description: >
             File status: AVAILABLE or INVALID (INVALID means that the file number is not in use, for example, a file in a tablespace that was dropped)
 
         - name: online_status
           type: keyword
+          dimension: true
```
Comment (ruflin) [inline packages/oracle/data_stream/tablespace/fields/fields.yml:52]: Does this have to be a dimension? It would be great to have a comment before each dimension field being specific on why something is a dimension. Something like:

```
# Dimension because of foo because otherwise bar
dimension: true
```

This will make sure anyone else modifying this later will know why something is a dimension.

File: packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json [hunk 5]
```diff
@@ -69,7 +69,7 @@
                                     "line_width": "2",
                                     "metrics": [
                                         {
-                                            "field": "oracle.performance.parse_count_total",
+                                            "field": "oracle.system_statistics.parse_count_total",
```
Comment (ruflin) [inline packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json:72]: How is this related to the PR?

File: packages/oracle/data_stream/memory/fields/base-fields.yml (hunk 2 above)
Comment (agithomas) [reply packages/oracle/data_stream/memory/fields/base-fields.yml:9]: I agree that it is not needed. A wrong change. I will revert

File: packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json (hunk 5 above)
Comment (agithomas) [reply packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json:72]: It is not related, it is an issue identified as part of the testing. I shall append these details in the changelog file regarding this issue fix.

Comment (ruflin) [reply packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json:72]: Or separate PR to get it out to the users quickly ;-)

File: packages/oracle/data_stream/performance/fields/fields.yml (hunk 3 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/fields/fields.yml:167]: There are 14 dimension fields including ecs fields.

File: packages/oracle/data_stream/tablespace/fields/fields.yml (hunk 4 above)
Comment (agithomas) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:52]: Perhaps, i was thinking too much when i chose the status as dimension fields. 

There exist no UNIQUE constraint i could find on this view and hence added the field. 

But, going with the fundamentals - can two files exist in a machine? No. 

But, can a similar situation arise if ASM is used especially with Alias? Not sure. I have to agree - those would be exceptions. 

I will go with the policy of adding dimension on `need basis` only and will not attempt `future proof`.   I will remove `status` and `online_status` from the list of dimensions.

File: packages/oracle/data_stream/memory/fields/agent.yml (hunk 1 above)
Comment (agithomas) [reply packages/oracle/data_stream/memory/fields/agent.yml]: Yes, the file was missing. Included the relevant fields in `ecs.yml`

File: packages/oracle/data_stream/performance/fields/fields.yml (hunk 3 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/fields/fields.yml:167]: It is 14 now, with the removal of two fields from the dimension list.

File: packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json (hunk 5 above)
Comment (agithomas) [reply packages/oracle/kibana/dashboard/oracle-b6b2c9f0-13a7-11ed-9607-2ba0819b3835.json]: Created a separate PR for the issue fix.

File: packages/oracle/data_stream/tablespace/fields/fields.yml (hunk 4 above)
Comment (ruflin) [reply packages/oracle/data_stream/tablespace/fields/fields.yml]: What will help with all these scenarios is adding as a comment to the fields.yml why something is a dimension.

File: packages/oracle/changelog.yml [hunk 6]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.11.0"
+  changes:
+    - description: Enabling changes to support TSDB enablement.
```
Comment (ruflin) [inline packages/oracle/changelog.yml:4]:
```suggestion
    - description: Add dimension fields to all dataset to support TSDB enablement. 
```

**Review by ruflin — APPROVED** [review]

File: packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -5,10 +5,6 @@ processors:
         field: sql.driver
         ignore_missing: true
         ignore_failure: true
-    - remove:
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:8]: Does this change also apply for other data_stream?

File: packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -128,6 +124,15 @@ processors:
         target_field: oracle.performance.wait.wait_class
         ignore_missing: true
         ignore_failure: true
+    - fingerprint:
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:127]: Same: does this change also applies for other data_streams?

File: packages/oracle/data_stream/performance/fields/fields.yml [hunk 9]
```diff
@@ -156,4 +164,5 @@
           description: Number of times waits of the class occurred for the session.
         - name: wait_class
           type: keyword
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/performance/fields/fields.yml:166]: Can we add a comment to describe why is this dimension?

File: packages/oracle/data_stream/system_statistics/fields/base-fields.yml [hunk 10]
```diff
@@ -6,6 +6,7 @@
   description: Data stream dataset.
 - name: data_stream.namespace
   type: constant_keyword
+  dimension: true
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/system_statistics/fields/base-fields.yml:9]: Is this wrongly added here? namespace is not a dimension, as discussed earlier.

File: packages/oracle/data_stream/performance/fields/fields.yml [hunk 11]
```diff
@@ -4,14 +4,22 @@
   fields:
     - name: machine
       type: keyword
+      dimension: true
       description: |
         Operating system machine name.
+    - name: query_id
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/performance/fields/fields.yml:10]: Does this apply to sysmetric & system_statistics?

File: packages/oracle/data_stream/tablespace/fields/fields.yml [hunk 12]
```diff
@@ -4,6 +4,7 @@
   fields:
```
Comment (lalit-satapathy) [inline packages/oracle/data_stream/tablespace/fields/fields.yml:4]: In this data_stream is _status/online_status_ dimension or metrics?

File: packages/oracle/data_stream/performance/fields/fields.yml (hunk 9 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/fields/fields.yml:166]: It is mentioned in the description of the field itself. I may not have anything more to add as a comment. 
As mentioned in the description 

 - Wait classes can be one of the following - Administrative, Application, Cluster, Commit, Concurrency, Configuration, Idle, Network, Other, Scheduler, System IO, User IO

File: packages/oracle/data_stream/performance/fields/fields.yml (hunk 11 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/fields/fields.yml:10]: No. It is applicable in `response_format: table` is used. Hence, not applicable for the said datastreams.

File: packages/oracle/data_stream/tablespace/fields/fields.yml (hunk 12 above)
Comment (agithomas) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: No, it is not.

File: packages/oracle/data_stream/system_statistics/fields/base-fields.yml (hunk 10 above)
Comment (agithomas) [reply packages/oracle/data_stream/system_statistics/fields/base-fields.yml:9]: Agree. I will correct it.

File: packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml (hunk 8 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:127]: No, only when a datastream has  `response_format: table`

File: packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:8]: No, only when a datastream has response_format: table

Comment (lalit-satapathy) [reply packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:8]: what about **tablespace**?

File: packages/oracle/data_stream/tablespace/fields/fields.yml (hunk 12 above)
Comment (lalit-satapathy) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: Not clear.. are we saying these two fields are neither dimension nor metrics?

File: packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (agithomas) [reply packages/oracle/data_stream/performance/elasticsearch/ingest_pipeline/default.yml:8]: Since there exist only one query in the `tablespace` datastream, it is not needed. 

The formula of -  whether to add the `fingerprint` logic for every `response_format: table` is subjective. It also depends on if there exist dimension fields that can uniquely determine time series. When such cases does not exist, it is when inventing new field such as `query_id` is necessary.  


In `performance` datastream,  there is inadequacy of dimension fields for all the queries configured and hence a `query_id` had to be invented.

File: packages/oracle/data_stream/tablespace/fields/fields.yml (hunk 12 above)
Comment (agithomas) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: Correct. These fields even if it is a `keyword` type  need not be a dimension.

Comment (agithomas) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: you can find more details here : https://github.com/elastic/integrations/pull/5307#discussion_r1111577596

Comment (agithomas) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: In an Integration usecase, in most cases, based on my experience, dimension fields are only those fields that uniquely define the `resource` or `asset` that are to be monitored. Not the states of the `resource`/ `asset`.  

Only in rare scenario  (`sql/metrics , response_format: table` ) `State` of an asset / resource may need to be included as `dimension`. Even in this case, it is needed, if other event fields are not sufficient enough to uniquely identify an event.

Comment (lalit-satapathy) [reply packages/oracle/data_stream/tablespace/fields/fields.yml:4]: Thanks to the pointer the the prior discussion.

File: packages/oracle/data_stream/performance/fields/fields.yml (hunk 11 above)
Comment (lalit-satapathy) [reply packages/oracle/data_stream/performance/fields/fields.yml:10]: ok.
